Pick the middle element as the pivot so that our quicksort doesn't have worst case performance for arrays that are already sorted.

mstange opened the pull request.

mstange commented:
This follow's @jimblandy's suggestion:
> @gregtatum Okay, yeah, this `partition` function really assumes that the pivot is the rightmost value in the range. So there's no reason for it to take a pivot parameter at all.
> That final `swap` just before the return is implicitly assuming that the pivot value is located at `right`.
> If you want to use arbitrary indices as pivot values, then you'll need to swap `pivot` with `right` at the top of `partition`.
> I have been nerd-sniped

gregtatum commented on src/content/data-table-utils.js:
```diff
@@ -64,7 +65,7 @@ export function sortDataTable<KeyColumnElementType>(
   function quickSort(left, right) {
     if (left < right) {
-      const pivot = right;
+      const pivot = (left + right) >> 1;
```
✨ MAGIC ✨

julienw replied:
maybe a comment to explain why this and why this works ? :)

jimblandy replied:
The comment would be something like:

QuickSort's effectiveness depends on its ability to partition the sequence being sorted into two subsequences of roughly equal length: by halving the length at each recursion level, the subproblems get smaller faster. To get a good partition, we must choose a pivot value that is close to the median of the values in the sequence: by definition, half of the values will fall before the median, half after it.

*If the sequence is already mostly sorted*, then choosing the first or last value as the pivot is probably as far from the median as you could possibly get; this is a "pessimal" choice. Choosing the middle value as the pivot is likely to be much closer to the median.

julienw replied:
+1 for this comment :) This is very pedagogic

mstange replied:
Thanks Jim, that's really a great comment! I'm putting it into the code verbatim.

mstange commented:
I've added some comments and made two tweaks: In `partition`, instead of swapping the pivot element to the end of the array at the start and then swapping it back into the `partitionIndex` position at the end, I've just stopped treating the pivot element specially. That means that the comparator function will now also be called to compare the pivot element to itself. But doing that seems to be quite a bit faster than the extra swaps.
And the more important optimization I added was to make `swap` check whether `i` and `j` are actually different. Oops.

mstange commented:
Oops, I introduced a mistake. The partition function no longer guarantees that pivotValue is at the partitionIndex on return, but the caller relies on that being the case. I'll fix it tomorrow, and I should probably add a test for it.
